File: contest/LeetCode_Weekly_Contest_206/Count_Unhappy_Friends.py
class Solution:
    def unhappyFriends(self, n: int, preferences: List[List[int]], pairs: List[List[int]]) -> int:
        res = set()
        for i, pair in enumerate(pairs):
            x, y = pair[0], pair[1]
            for j, pair2 in enumerate(pairs[i + 1:]):
                u, v = pair2[0], pair2[1]
                if preferences[x].index(u) < preferences[x].index(y) \
                    and preferences[u].index(x) < preferences[u].index(v):
                    res.add(x)
                    res.add(u)
                u, v = v, u
                if preferences[x].index(u) < preferences[x].index(y) \
                    and preferences[u].index(x) < preferences[u].index(v):
                    res.add(x)
                    res.add(u)

            x, y = y, x
            for j, pair2 in enumerate(pairs[i + 1:]):
                u, v = pair2[0], pair2[1]
                if preferences[x].index(u) < preferences[x].index(y) \
                    and preferences[u].index(x) < preferences[u].index(v):
                    res.add(x)
                    res.add(u)
                u, v = v, u
                if preferences[x].index(u) < preferences[x].index(y) \
                    and preferences[u].index(x) < preferences[u].index(v):
                    res.add(x)
                    res.add(u)
        return len(res)

File: contest/LeetCode_Weekly_Contest_206/Count_Unhappy_Friends.cpp
class Solution {
public:
    unordered_map<int, int> m[1000];
    bool work(int x, int y, int u, int v) {
        int pu = m[x][u];
        int py = m[x][y];
            
        int px = m[u][x];
        int pv = m[u][v];
        
        if (pu < py && px < pv) return true;
        return false;
    }
    int unhappyFriends(int n, vector<vector<int>>& ps, vector<vector<int>>& pairs) {
        vector<bool> vis(n, true);
        
        for (int i = 0; i < n; ++i) {
            for (int k = 0; k < ps[i].size(); ++k)
                m[i][ps[i][k]] = k;
        }
        for (int i = 0; i < pairs.size(); ++i) {
            for (int j = 0; j < pairs.size(); ++j) if (i != j) {
                int x = pairs[i][0], y = pairs[i][1];
                int u = pairs[j][0], v = pairs[j][1];
                
                if (work(x, y, u, v)) vis[x] = false;
                if (work(y, x, u, v)) vis[y] = false;
                if (work(x, y, v, u)) vis[x] = false;
                if (work(y, x, v, u)) vis[y] = false;
            }
        }
        int ret = 0;
        for (int i = 0; i < n; ++i)
            if (!vis[i]) ++ret;
        return ret;
    }
};
